### PR TITLE
Add mechanism for blocking oneDNN cache clearing

### PR DESCRIPTION
**Pull request by grygielski**
### PR types
New features

### PR changes
Others

### Describe
This PR adds a mechanism for preventing oneDNN cache clearing without changes in Executor. It solves problems mentioned in https://github.com/PaddlePaddle/Paddle/issues/25988.

### CODE REVIEW

**Comment by grygielski** [comment]
@luotao1 I've prepared this PR as a preview to my solution for conditionally clearing oneDNN cache. It's applied for now just in `run_program_op` but can be extended to others like `conditional_block_op` and so on. Let me know what do you think about this solution.

File: paddle/fluid/operators/run_program_op.h [hunk 1]
```diff
@@ -262,6 +265,9 @@ class RunProgramOpKernel : public framework::OpKernel<T> {
     }
     VLOG(2) << "The number of sub scopes after forward: "
             << out_scope_vec->front()->kids().size();
+#ifdef PADDLE_WITH_MKLDNN
+    DontClearMKLDNNCache(ctx.GetPlace());
+#endif
```
Comment (luotao1) [inline paddle/fluid/operators/run_program_op.h:272]:
```
#ifdef PADDLE_WITH_MKLDNN
  if(FLAGS_use_mkldnn)
      DontClearMKLDNNCache(ctx.GetPlace());
#endif
```

**Review by luotao1 — COMMENTED** [review]
> It's applied for now just in run_program_op but can be extended to others like conditional_block_op and so on

I think it's OK,   @zhhsplendid @chenwhql Could you help review it?

File: paddle/fluid/platform/device_context.h [hunk 2]
```diff
@@ -506,6 +509,7 @@ class MKLDNNDeviceContext : public CPUDeviceContext {
   mkldnn::engine engine_;
   std::shared_ptr<BlobMap> p_blobmap_;
   std::shared_ptr<std::mutex> p_mutex_;
+  bool block_next_cache_clearing = false;
```
Comment (zhhsplendid) [inline paddle/fluid/platform/device_context.h]: Our private members are named with a trailing underscore, like line 509 - line 512.

Could you rename it to "block_next_cache_clearing_" ?

**Review by luotao1 — APPROVED** [review]

**Review by wojtuss — APPROVED** [review]
LGTM

**Pull request merged**